File: app/web/src/data/FunctionsDs.ts
import { getFunctions, httpsCallable } from "firebase/functions"
import firebaseApp from "../utility/FirebaseApp";
import { AppError } from "./Model/AppError";
import type { SetUsernameStatus } from "./Model/SetUsernameStatus";

interface FunctionsResponse<S, D> {
    status: S;
    data: D;
}

export default class FunctionsDs {
    private static functions = getFunctions(firebaseApp())

    private static async callFunction<T>(name: string, params?: Record<string, any>): Promise<T> {
        try {
            
            const callable = httpsCallable(this.functions, name);
            const result = await callable(params);

            if (typeof result.data === 'string') {
                const data = JSON.parse(result.data);
                return data as T;
            }
            throw new Error('Invalid response format');
        } catch (error) {
            console.error(error);
            throw AppError.InvalidResponse
        }
    }

    // static async acceptInvite(invitePath: string): Promise<FunctionsResponse<InviteStatus, 'AcceptInviteData'>> {
    //     return this.callFunction('acceptInvite', { invitePath });
    // }

    // static async sendInvite(uid: string, difficulty: string): Promise<FunctionsResponse<InviteStatus, 'SendInviteData'>> {
    //     return this.callFunction('sendInvite', { invitee: uid, difficulty });
    // }

    static async setUsername(username: string): Promise<FunctionsResponse<SetUsernameStatus, never>> {
        return this.callFunction('setUsername', { username });
    }

    // static async verifyDuelBoard(duelPath: string): Promise<FunctionsResponse<VerifyDuelBoardStatus, never>> {
    //     return this.callFunction('verifyDuelBoard', { duelPath });
    // }

    // static async requestMatchmaking(): Promise<FunctionsResponse<MatchmakingStatus, 'MatchmakingData'>> {
    //     return this.callFunction('matchmaking');
    // }

    // static async deleteAccount(): Promise<FunctionsResponse<DeleteAccountStatus, 'MatchmakingData'>> {
    //     return this.callFunction('deleteAccount');
    // }
}

File: app/ios/Sudoku Battles/Data/FunctionsDs.swift
import FirebaseFunctions
import FirebaseAuth
import FirebaseFirestore

class FunctionsDs {
    static let shared = FunctionsDs()
    private let functions: Functions

    init() {
        self.functions = Functions.functions()
        if Bundle.main.dev {
            self.functions.useEmulator(withHost: "\(Bundle.main.firebaseHost)", port: 5001)
        }
    }

    private func callFunction<T: Decodable>(_ name: String, params: [String: Any]? = nil) async throws -> T {
        return try await withCheckedThrowingContinuation { continuation in
            functions.httpsCallable(name).call(params) { result, error in
                if let error = error {
                    logger.error("\(error)")
                    continuation.resume(throwing: error)
                } else {
                    guard let data = (result?.data as? String)?.data(using: .utf8) else {
                        continuation.resume(throwing: AppError.networkError)
                        return
                    }
                    do {
                        let response = try JSONDecoder().decode(T.self, from: data)
                        continuation.resume(returning: response)
                    } catch {
                        logger.error("\(error)")
                        continuation.resume(throwing: AppError.invalidResponse)
                    }
                }
            }
        }
    }

    func acceptInvite(invitePath: String) async throws -> FunctionsResponse<InviteStatus, AcceptInviteData> {
        try await callFunction("acceptInvite", params: ["invitePath": invitePath])
    }
    func sendInvite(uid: String, difficulty: Difficulty) async throws -> FunctionsResponse<InviteStatus, SendInviteData> {
        try await callFunction("sendInvite", params: ["invitee": uid, "difficulty": difficulty.rawValue])
    }
    func setUsername(username: String) async throws -> FunctionsResponse<SetUsernameStatus, Never> {
        try await callFunction("setUsername", params: ["username": username])
    }
    func verifyDuelBoard(duelPath: String) async throws -> FunctionsResponse<VerifyDuelBoardStatus, Never> {
        try await callFunction("verifyDuelBoard", params: ["duelPath": duelPath])
    }
    func requestMatchmaking() async throws -> FunctionsResponse<MatchmakingStatus, MatchmakingData> {
        try await callFunction("matchmaking")
    }
    func deleteAccount() async throws -> FunctionsResponse<DeleteAccountStatus, MatchmakingData> {
        try await callFunction("deleteAccount")
    }
}

